File: of-core-mods/patch_refiner.pl
#!/usr/bin/perl
use strict;
use warnings;

my $root = "$ENV{HOME}/openfoam/RGP-13/of-core-mods/fvMeshTopoChangers";
my $cpath = "$root/refiner/refiner_fvMeshTopoChanger.C";
my $mpath = "$root/Make/files";

# ---- 1) Retarget Make/files LIB to FOAM_USER_LIBBIN ----
{
    local $/; open my $fh, '<', $mpath or die "open $mpath: $!";
    my $m = <$fh>; close $fh;
    my $old = 'LIB = $(FOAM_LIBBIN)/libfvMeshTopoChangers';
    my $new = 'LIB = $(FOAM_USER_LIBBIN)/libfvMeshTopoChangers';
    die "Make/files LIB line not found\n" unless index($m, $old) >= 0;
    $m =~ s/\Q$old\E/$new/;
    open my $out, '>', $mpath or die; print $out $m; close $out;
    print "Make/files patched -> FOAM_USER_LIBBIN\n";
}

# ---- 2) Patch the .C ----
my $c;
{ local $/; open my $fh, '<', $cpath or die "open $cpath: $!"; $c = <$fh>; close $fh; }

# 2a) add include
my $inc_anchor = '#include "addToRunTimeSelectionTable.H"';
die "include anchor not found\n" unless index($c, $inc_anchor) >= 0;
die "include already present\n" if index($c, '#include "polyDistributionMap.H"') >= 0;
$c =~ s/\Q$inc_anchor\E/$inc_anchor\n#include "polyDistributionMap.H"/;

# 2b) replace distribute body
my $old_block = <<'END_OLD';
void Foam::fvMeshTopoChangers::refiner::distribute
(
    const polyDistributionMap& map
)
{
    // Redistribute the mesh cutting engine
    meshCutter_.distribute(map);
}
END_OLD

my $new_block = <<'END_NEW';
void Foam::fvMeshTopoChangers::refiner::distribute
(
    const polyDistributionMap& map
)
{
    // Redistribute the mesh cutting engine (cellLevel/pointLevel/history)
    meshCutter_.distribute(map);

    // Redistribute the protected-cell marking to the new decomposition.
    // Upstream refiner::distribute only handles meshCutter_; protectedCells_
    // is sized to the OLD local cell count and would be stale (wrong size /
    // mis-indexed) after redistribution, corrupting the next (un)refinement.
    if (protectedCells_.size())
    {
        labelList protectedCell(protectedCells_.size());
        forAll(protectedCell, celli)
        {
            protectedCell[celli] = protectedCells_.get(celli);
        }

        map.distributeCellData(protectedCell);

        PackedBoolList newProtectedCell(protectedCell.size());
        forAll(protectedCell, celli)
        {
            newProtectedCell.set(celli, protectedCell[celli]);
        }
        protectedCells_.transfer(newProtectedCell);
    }
}
END_NEW

die "distribute() block not matched exactly\n" unless index($c, $old_block) >= 0;
my $cnt = ($c =~ s/\Q$old_block\E/$new_block/);
die "expected 1 replacement, got $cnt\n" unless $cnt == 1;

open my $out, '>', $cpath or die; print $out $c; close $out;
print "refiner_fvMeshTopoChanger.C patched: include + protectedCells_ redistribution\n";

File: of-core-mods/patch_refiner.py
#!/usr/bin/env python3
import os, sys

root = os.path.expanduser("~/openfoam/RGP-13/of-core-mods/fvMeshTopoChangers")
cpath = os.path.join(root, "refiner", "refiner_fvMeshTopoChanger.C")
mpath = os.path.join(root, "Make", "files")

# 1) Retarget Make/files LIB to FOAM_USER_LIBBIN
with open(mpath) as f:
    m = f.read()
old_lib = "LIB = $(FOAM_LIBBIN)/libfvMeshTopoChangers"
new_lib = "LIB = $(FOAM_USER_LIBBIN)/libfvMeshTopoChangers"
assert old_lib in m, "Make/files LIB line not found"
m = m.replace(old_lib, new_lib)
with open(mpath, "w") as f:
    f.write(m)
print("Make/files patched -> FOAM_USER_LIBBIN")

# 2) Patch the .C
with open(cpath) as f:
    c = f.read()

# 2a) Add polyDistributionMap.H include (for template distributeCellData)
inc_anchor = '#include "addToRunTimeSelectionTable.H"'
inc_add = inc_anchor + '\n#include "polyDistributionMap.H"'
assert inc_anchor in c, "include anchor not found"
assert '#include "polyDistributionMap.H"' not in c, "include already present"
c = c.replace(inc_anchor, inc_add, 1)

# 2b) Replace the distribute() body
old_block = (
    "void Foam::fvMeshTopoChangers::refiner::distribute\n"
    "(\n"
    "    const polyDistributionMap& map\n"
    ")\n"
    "{\n"
    "    // Redistribute the mesh cutting engine\n"
    "    meshCutter_.distribute(map);\n"
    "}\n"
)
new_block = (
    "void Foam::fvMeshTopoChangers::refiner::distribute\n"
    "(\n"
    "    const polyDistributionMap& map\n"
    ")\n"
    "{\n"
    "    // Redistribute the mesh cutting engine (cellLevel/pointLevel/history)\n"
    "    meshCutter_.distribute(map);\n"
    "\n"
    "    // Redistribute the protected-cell marking to the new decomposition.\n"
    "    // Upstream refiner::distribute only handles meshCutter_; protectedCells_\n"
    "    // is sized to the OLD local cell count and would be stale (wrong size /\n"
    "    // mis-indexed) after redistribution, corrupting the next (un)refinement.\n"
    "    if (protectedCells_.size())\n"
    "    {\n"
    "        labelList protectedCell(protectedCells_.size());\n"
    "        forAll(protectedCell, celli)\n"
    "        {\n"
    "            protectedCell[celli] = protectedCells_.get(celli);\n"
    "        }\n"
    "\n"
    "        map.distributeCellData(protectedCell);\n"
    "\n"
    "        PackedBoolList newProtectedCell(protectedCell.size());\n"
    "        forAll(protectedCell, celli)\n"
    "        {\n"
    "            newProtectedCell.set(celli, protectedCell[celli]);\n"
    "        }\n"
    "        protectedCells_.transfer(newProtectedCell);\n"
    "    }\n"
    "}\n"
)
assert old_block in c, "distribute() block not matched exactly"
c = c.replace(old_block, new_block, 1)

with open(cpath, "w") as f:
    f.write(c)
print("refiner_fvMeshTopoChanger.C patched: include + protectedCells_ redistribution")
